### PR TITLE
King castle

### PR DESCRIPTION
**Pull request by MarciaOrr**
Please review. I don't like overriding Piece::valid_move? just to avoid calling Piece::move_obstruction? after move_legal?. I tried numerous ways to get around this including a King::move_obstruction?. That solution was worse because it had to know if it was a castle move or a regular move - and I ended up with an instance variable. I must to call Piece:move_obstruction? when doing a castle. 

I hate lines 38 - 42 in legal_rook_move? but Rubocop insisted on this syntax. I feel like I spin around on my head three times in order to make Rubocop happy when it is only one additional line in a method - but the code (to me) actually reads cleaner my original way. 

I did a re-work in piece.rb. I created an update_piece method that updates_attributes for the instance of a piece class (for me a King or a Rook) and calls create_move!. I tested it with a pawn on the chessboard. The pawn tests still pass. With this change, I'm also passing a string into create_move for the move_type. What I can't figure out is how to get 'castled' into the move for the king. It is going in for the rook. A captured piece will have 'captured' in the move_type and the piece performing the capture will have 'capturing'.

### CODE REVIEW

**Comment by jimxlin** [comment]
I also had to refactor some of `piece.rb` in my PR.. We will have to figure out how to merge our two changes to the file.

File: app/models/king.rb
Comment (jimxlin) [inline app/models/king.rb]: A ternary operator would work here: `new_rook_col = rook.col.zero? ? 3 : 5`

File: app/models/king.rb
Comment (jimxlin) [inline app/models/king.rb]: The rook also must not have moved to be eligible for castling.

File: app/models/king.rb
Comment (jimxlin) [inline app/models/king.rb]: In castling, the rook cannot be in the same column of the king. You would need to call `legal_rook_move?(to_col + 1)`, or `to_col - 1`, depending on which way the king is castling.

Comment (MarciaOrr) [reply app/models/king.rb]: The king's to_col in the call to get_rook? is used to determine if the castle is kingside or queenside. Within get_rook? that column is used to get either the rook in column 0 or the rook in column 7 (provided that a rook piece exists at that square and has not moved).

File: app/models/king.rb
Comment (MarciaOrr) [reply app/models/king.rb]: I thought that I checked that in rook_at at line 49. There is also a test where the rook has a move and prevents the castle.

File: app/models/king.rb
Comment (MarciaOrr) [reply app/models/king.rb]: Done. Thanks.

File: app/models/king.rb
Comment (jimxlin) [reply app/models/king.rb]: It's possible that a rook moves, then moves back to its starting position.

File: app/models/king.rb
Comment (jimxlin) [reply app/models/king.rb]: I see, I misunderstood how the method works.

**Comment by jimxlin** [comment]
On creating moves: Only one Move should be created per move, e.g. after a capture, the captor should create a Move, but the captive does not. For castling, I suggest only having the King create the Move, as that is the piece that the player moves.

On castling: did you implement the logic that performs the capture? I can't seem to find it.

File: app/models/king.rb
Comment (MarciaOrr) [reply app/models/king.rb]: That is true - a rook can move all over the board and return to its original position.

The Moves table contains all of the moves. The method rook_at queries that table as the check that the rook has not moved (line 49). The rook's position on the board is not used to determine that it has never, ever moved. The method only makes use of the  position to determine if a rook even exists at 0,0 || 0,7 || 7,0 || 7,7. 

I have manually tested this concept with IRB and the web application. I moved the rook at 7,7 to a location [3,7] and moved it back to[ 7,7]. I perform a page refresh between every move to ensure that the move happened. Then I manually queried the table for Rook.find(90).moves.empty?. It returned false - because the rook moved. Next I did Move.delete_all and the query again. It returned true to indicate that the rook had not ever moved.

king_spec.rb contains this test as the last one. A move is explicitly created for the rook. In the previous test, a move is created for the king.) Next move_to! is called for the king. The move cannot occur since the rook has already moved.

In order to castle, the method rook_at does enforce the rule that a castle can only occur when it is the first move for both the king and the rook.

**Comment by MarciaOrr** [comment]
There is no capture in a castle. Both the king and the rook MOVE to a new position on the board. In my opinion, the king and the rook should each have a record in the Move table because their to_position has changed.

I think that a captured piece should have a final move in the Move table. It's to_position has changed. In live chess, the piece leaves the board - it moves. We can discuss this point as a group. I can see where there is only one move  that results in a capture.

Piece.rb contains code in this PR to complete the move_type field for some Move record. For a captured piece it contains the to_position of [-1,-1] and the string 'captured' for the move_type. (Line 43 of piece.rb within capture_piece?. The capturing piece also has a Move record. The move_type is set to 'capturing' on line 23. It is set to 'moved' otherwise.

**Comment by jimxlin** [comment]
^ Sorry for the typo, I meant that I could not find the logic that moves the rook during castling.

File: app/models/king.rb
Comment (jimxlin) [reply app/models/king.rb]: Ah, I did not look at line 53 close enough, sorry for the mistake.

**Comment by MarciaOrr** [comment]
king.rb. Line 39. rook.update_piece(row, new_rook_col, 'castled')
piece.rb. Line 35. new method update_piece(to_row, to_col, state)
   Line 38. update_attributes is called for the instance of the piece. Here it is the rook.
   Line 39. adds the rook's move to the Moves table.

   Yes. the rook moves first - but one of the pieces has to move first.
This is OK because:
   1) the king has moved in accordance with the Castle rules (legal_king_move?)
   2) A rook is in the correct corner, it has not moved (king.rb - lines 35)
   3) the path between the king and the rook is clear (line 37)

For the king - 
  move_legal? king.rb - Line 11 returns true to valid_move? (king.rb line 2) which returns true to move_to! (piece.rb - line 18.)
   Within move_to!, capture_piece? is called because this is the parent method for all pieces. Since there is no capture, update_piece on line 25 is called - which updates the row and column for this instance of a Piece which is the king. create_move! is also called.

**Comment by jimxlin** [comment]
Ok I see it on line 43. Sorry, its a bit difficult for me to follow the logic when many methods are involved (which I understand is necessary for this kind of move).

File: app/models/king.rb
Comment (jimxlin) [inline app/models/king.rb]: It would be more convenient to just do `rook.update_attributes(...)`, but again, we should discuss whether or not castling should be considered a single move.

**Comment by MarciaOrr** [comment]
I was just taking a quick look at the check/checkmate/stalemate PR. It affects this PR.

It is necessary to ensure that the king did not move itself into one of these states BEFORE the rook moves. So legal_king_move? will have to ascertain if this is the case by calling whatever method you will use in your PR. Then if the king is in a threatened state, legal_king_move? will return false and move_legal? will return false, and valid_move? will return false - so neither the king nor the rook will move. I think that your front-end successfully takes care of popping the king back into position on the board. The rook never moved because legal_rook_move? is not called.

I just want to make sure that we both remember this important point as our code continues forward.

File: app/models/king.rb
Comment (MarciaOrr) [reply app/models/king.rb]: If the team decides that a castle is a single move that only belongs to the king - then yes.

**Comment by jimxlin** [comment]
I think it's best to first review and merge my check/checkmate/stalemate PR, as the castling logic will need to check if the king is in check. We can then work on how best to integrate this PR with that one before merging this one.

**Comment by MarciaOrr** [comment]
I like your comment on so many methods and difficulty in following the logic. I completely agree. 

Actually it's rubocop that forced me to break it up into so many methods because of the 10 line maximum. I had a longer, single method for the rook that to me was pretty concise and it was easier to follow the logic flow. (The reviewer wasn't jumping all around in the file.) Rubocop also complained about complexity. 

This is a Jacob teaching opportunity to the team or a decision point as the Scrum master. I think there should be a single method to check the king. (which we have). If that passes, a single method to check the rook rules and update its position.

File: app/models/king.rb
Comment (jacobwgillespie) [inline app/models/king.rb]: Once #41 merges there will likely be some merge conflicts, etc., but here's something that might simplify this - by overriding the `valid_move?` method in the king, you have to keep track of how the `valid_move?` in piece changes, so if you add something there thinking it'll apply to all the pieces, you'd also have to add it to the king, etc., and the relationship might not be super obvious.

I think what you're doing is saying that a king is never obstructed (since you removed that check from the copied method), so rather than overriding `valid_move?`, it's probably better to just override `move_obstructed?` like so:

```ruby
def move_obstructed?(to_row, to_col)
  false
end
```

That way it's clearer that you're saying a king can't be obstructed.

File: app/models/king.rb
Comment (jacobwgillespie) [inline app/models/king.rb]: Can you name this with a `?` at the end, to make it clear that this returns `true` / `false`?

File: app/models/king.rb
Comment (jacobwgillespie) [inline app/models/king.rb]: This method can be simplified:

```ruby
def legal_castle_move?(to_row, to_col)
  legal_king_move?(to_row, to_col) && legal_rook_move?(to_col)
end
```

In fact any time you see some variant of using if / else to return true or false, you can usually just return the condition:

```ruby
return true if something
return false

# becomes:

return something
```

```ruby
if something
  return true
else
  return false
end

# becomes:

return something
```

```ruby
if something
  if other_something
    return true
  else
    return false
  end
else
  return true
end

# becomes:

return something ? other_something : true
```

Or something like that.  The `something` and `other_something` can be full expressions (`asdf != 123`) or method calls or whatever.

The exception is cases like our `valid_move?` where there are a lot of conditions, so a simple fall-through approach of "guard clauses" is more readable than deciphering `&&`s and `||`s.

File: app/models/king.rb
Comment (jacobwgillespie) [inline app/models/king.rb]: This can also have a `return false if row_diff != 0` because at this point the only remaining valid move is castling and you can't change rows on a castle move.

That also means you don't need to pass `to_row` to the `legal_castle_move` method anymore.

File: app/models/king.rb
Comment (jacobwgillespie) [inline app/models/king.rb]: Can this be named `can_castle_to?(to_col)`?

(because then see below about `can_castle_to?` for the rook)

File: app/models/king.rb
Comment (jacobwgillespie) [inline app/models/king.rb]: This can be removed if you add the check above to `move_legal?`

File: app/models/king.rb
Comment (jacobwgillespie) [inline app/models/king.rb]: Can you move this into the Rook class and name it `can_castle_to?(to_col)`?

It is essentially logic for the rook, so it makes sense to have it there vs inside the King class.  This would change your `legal_castle_move?(to_col)` method to something like:

```ruby
def legal_castle_move?(to_col)
  rook = get_rook(to_col)
  rook && rook.can_castle_to?(to_col) && can_castle_to?(to_col)
end
```

File: app/models/king.rb
Comment (jacobwgillespie) [inline app/models/king.rb]: `get_rook?` probably shouldn't have the `?` at the end of the method, because it's not a boolean method (it returns nil or an object).

Also, `get_rook?` and `rook_at` are named very similar names, and they could actually be combined (naming the combined method `get_castling_rook` here, but feel free to do w/e make sense to you):

```ruby
def get_castling_rook(to_col)
  rook_col = col < to_col ? 7 : 0
  rook = game.piece_at(row, rook_col)
  rook if rook && rook.moves.empty?
end
```

File: app/models/king.rb
Comment (jacobwgillespie) [inline app/models/king.rb]: This `update_piece` should probably be moved into its own method so that the check methods (the ones ending in `?`) just answer questions about *if* it's okay, and then a method like `castle!(to_col)` actually does the castling.

```ruby
# NOTE: the king will be moved through the normal move_to!
def castle!(to_col)
  rook = get_castling_rook(to_col)
  return unless rook
  new_rook_col = rook.col.zero? ? 3 : 5
  rook.update_piece(row, new_rook_col, 'castled')
end
```

You could add a method named `is_castling?(to_row, to_col)` to the Piece class:

```ruby
def is_castling?(to_row, to_col)
  false
end
```

Then override it in the King class:

```ruby
def is_castling?(to_row, to_col)
  row == to_row && (col - to_col).abs == 2
end
```

Then inside `move_to!` in the Piece class, right after `capture_piece`, you could have:

```ruby
# ...
capture_piece(to_row, to_col)
castle!(to_col) if is_castling?(to_row, to_col)
# ...
```

That would take care of moving the rook.

Just one idea for how to separate out the validation from the actual move.

**Review by jacobwgillespie — CHANGES_REQUESTED** [review]
Looking good!  A few comments / suggestions - let me know if you have any questions or if I'm missing anything 👍

File: app/models/king.rb
Comment (MarciaOrr) [inline app/models/king.rb]: Yes. I had tried that since a regular king move is never obstructed since can only move to an adjacent square.

Comment (MarciaOrr) [reply app/models/king.rb]: I have to verify that the path between the king's current square and the rook's current square is not obstructed.

At one point, I did have the above move_obstructed? for a non-castle move. Then I realized that I had to call Piece::move_obstructed. I just couldn't figure out how to do that from the King class.

Comment (jimxlin) [reply app/models/king.rb]: A possible alternative is to create a method just for checking if the path between the castlers is clear. This method could also make sure that the squares in the path cannot be attacked. Something like:
```
def path_is_clear?(castle_path)
  # this is mostly pseudo-code
  castle_path.each {|square| return false if square_is_empty && square_is_not_attackable}
  true
end
```

File: app/models/rook.rb
Comment (jimxlin) [inline app/models/rook.rb]: I suggested a path_is_clear method because we still have to implement the castle rule where the squares along the path of the king cannot be valid moves of enemy pieces. A path_is_clear method could check for that, and for the presence of pieces between the king and rook as well.

Comment (MarciaOrr) [inline app/models/rook.rb]: The squares along the path between the rook and the king do not matter. The enemy can't attack until the king and the rook are in their new squares.

Comment (MarciaOrr) [reply app/models/rook.rb]: The chess rule is no pieces between the king's original location and the rook's original location. Calling move_obstructed? enforces that rule.

The path between the rook and the king during a castle doesn't matter - just like the path that any piece takes doesn't matter until it is in its new square. What matters is the final square for the piece and the question, "Did moving to this square place my king in check?" If I'm the King, "Did I put myself into check?"

What does matter first is the king's move. By moving 2 squares horizontally - did the king put itself into check? Yes - move it back.

Now move the rook. Did this move place the king in check? If yes - move it AND the king back.

Comment (jimxlin) [reply app/models/rook.rb]: There is an additional rule: https://en.wikipedia.org/wiki/Castling
See rule 5 in the Requirements section.

File: app/models/king.rb
Comment (jacobwgillespie) [inline app/models/king.rb]: Looks good!  You can remove the `# Jacob` comments ;)

File: app/models/pawn.rb
Comment (jacobwgillespie) [inline app/models/pawn.rb]: This probably shouldn't have the `?` (the parent is just `capture_piece`)

File: app/models/rook.rb
Comment (jacobwgillespie) [reply app/models/rook.rb]: Let's add the `path_is_clear` in a followup PR, to integrate it with the check logic.

@MarciaOrr it matters here because one of the castling rules is a king cannot move *through* check while castling, so this is the one case where the in-between spaces are important.  But I think it'll be cleaner to open a new PR for that integration after this PR is merged.

**Review by jacobwgillespie — DISMISSED** [review]
One `?` to remove, and you can remove the PR feedback comments, but other than that this looks good to go!  🚀

**Comment by jimxlin** [comment]
As mentioned in a previous discussion with Marcia, `piece.rb` has some fairly complex conflicts with the check/checkmate/stalemate PR that was just merged. I think this warrants a group discussion on how to resolve the conflicts.

**Comment by jimxlin** [comment]
LGTM :rocket: 
The test for `should not allow king castle: next square under attack` looks good, I also test-ran that feature on the front-end sample board, and it worked.

The latest changes in master don't touch back end logic, so this merge should go smoothly.

As Jacob is no longer available to approve PRs, and you did implement his requested changes, I dismissed his requests.

**Comment by MarciaOrr** [comment]
Thank Jim for the review. I did my initial code develop by playing the Test Game. I had the king in and out of check at the destination as well as in the intermediate square. I would watch the King pop back into its original square. I also tested with an obstructed path between the king and the rook.My only concern was the instance variable that I created to hold the ID of the castling rook. That was the only way that I could think of to access it for the move_type string similar to your capture string.

**Comment by MarciaOrr** [comment]
Thank Jim for the review. I did my initial code develop by playing the Test Game. I had the king in and out of check at the destination as well as in the intermediate square. I would watch the King pop back into its original square. I also tested with an obstructed path between the king and the rook. At the last minute, I wrote a test to make sure that the rook did move as well.

My only concern was the instance variable that I created to hold the ID of the castling rook. That was the only way that I could think of to access it for the move_type string similar to your capture string. Please take a look at that for me since we don't have Jacob. If it is good to go, I will merge it into master tomorrow.

File: app/models/king.rb
Comment (jacobwgillespie) [inline app/models/king.rb]: Instance variables are automatically nil and don't need to be declared, so you don't need this here (kinda weird, but that's how it is)

File: app/models/king.rb
Comment (jimxlin) [inline app/models/king.rb]: The use of an instance variable is okay. But it's not necessary to record the rook's id; we wouldn't be using it in any logic. Usually the castling direction is identified by 'kingside' (to_col == 6) or 'queenside' (to_col == 2).

Comment (jacobwgillespie) [inline app/models/king.rb]: You might be able to decrease the instance variable's scope by using [memoization](http://gavinmiller.io/2013/basics-of-ruby-memoization/).

```ruby
def castling?
  return "castle. rook: #{get_castling_rook(to_col).id}" if row == to_row && (col - to_col).abs == 2
  "normal"
end

def get_castling_rook(to_col)
  @_castling_rook ||= {}
  @_castling_rook[to_col] ||= begin
    rook_col = col < to_col ? 7 : 0
    rook = game.piece_at(row, rook_col)
    return false unless rook && rook.moves.empty?
    return rook if rook.king_rook_path_clear?(col)
    false
  end
end
```

The first line `@_castling_rook ||= {}` sets `@_castling_rook` to an empty hash if it is currently `nil`, so it can store cached pieces based on the key value of `to_col`.

The second either returns what's currently at `@_castling_rook[to_col]`, or it calculates it inside the block, saves, the value, then returns it.

In that way, you can call the method again later and get the same rook as before.  And you don't need to reference the instance variable in more than one method.

**Review by jacobwgillespie — COMMENTED** [review]
Still around, and happy to keep reviewing stuff as able.  🙂  Just a few comments on how to minimize the instance variable.

File: app/models/king.rb
Comment (MarciaOrr) [reply app/models/king.rb]: I like that solution and implemented it. Totally gets rid of the instance variable.

**Comment by jimxlin** [comment]
LGTM

**Pull request merged**
